packages/grafana-flamegraph/src/FlameGraph/FlameGraphCanvas.tsx

```diff
@@ -0,0 +1,259 @@
+        const item = convertPixelCoordinatesToBarCoordinates(
+          { x: e.nativeEvent.offsetX, y: e.nativeEvent.offsetY },
+          root,
+          direction,
+          depth,
+          pixelsPerTick,
+          totalViewTicks,
+          rangeMin
+        );
+
+        if (item) {
+          setMousePosition({ x: e.clientX, y: e.clientY });
+          setTooltipItem(item);
+        }
+      }
+    },
+    [rangeMin, rangeMax, totalViewTicks, clickedItemData, setMousePosition, root, direction, depth]
+  );
+
+  const onGraphMouseLeave = useCallback(() => {
+    setTooltipItem(undefined);
+  }, []);
+
+  // hide context menu if outside the flame graph canvas is clicked
+  useEffect(() => {
+    const handleOnClick = (e: MouseEvent) => {
+      if (
+        e.target instanceof HTMLElement &&
+        e.target.parentElement?.id !== 'flameGraphCanvasContainer_clickOutsideCheck'
+      ) {
+        setClickedItemData(undefined);
+      }
+    };
+    window.addEventListener('click', handleOnClick);
+    return () => window.removeEventListener('click', handleOnClick);
+  }, [setClickedItemData]);
+
+  return (
+    <div className={styles.graph}>
+      <div className={styles.canvasWrapper} id="flameGraphCanvasContainer_clickOutsideCheck" ref={sizeRef}>
+        <canvas
+          ref={graphRef}
+          data-testid="flameGraph"
+          onClick={onGraphClick}
+          onMouseMove={onGraphMouseMove}
+          onMouseLeave={onGraphMouseLeave}
+        />
+      </div>
+      <FlameGraphTooltip position={mousePosition} item={tooltipItem} data={data} totalTicks={totalViewTicks} />
+      {clickedItemData && (
+        <FlameGraphContextMenu
+          itemData={clickedItemData}
+          onMenuItemClick={() => {
+            setClickedItemData(undefined);
+          }}
+          onItemFocus={() => {
+            setRangeMin(clickedItemData.item.start / totalViewTicks);
+            setRangeMax((clickedItemData.item.start + clickedItemData.item.value) / totalViewTicks);
+            onItemFocused(clickedItemData);
+          }}
+          onSandwich={() => {
+            onSandwich(data.getLabel(clickedItemData.item.itemIndexes[0]));
+          }}
+        />
+      )}
+    </div>
+  );
+};
+
+const getStyles = () => ({
+  graph: css({
+    label: 'graph',
+    overflow: 'auto',
+    height: '100%',
+    flexGrow: 1,
+    flexBasis: '50%',
+  }),
+  canvasContainer: css({
+    label: 'canvasContainer',
+    display: 'flex',
+  }),
+  canvasWrapper: css({
+    label: 'canvasWrapper',
+    cursor: 'pointer',
+    flex: 1,
+    overflow: 'hidden',
+  }),
+  sandwichMarker: css({
+    label: 'sandwichMarker',
+    writingMode: 'vertical-lr',
+    transform: 'rotate(180deg)',
+    overflow: 'hidden',
+    whiteSpace: 'nowrap',
+  }),
+  sandwichMarkerIcon: css({
+    label: 'sandwichMarkerIcon',
+    verticalAlign: 'baseline',
+  }),
+});
+
+const convertPixelCoordinatesToBarCoordinates = (
```

Comment:
One of the main changes. We have to do a tree search for the right node that corresponds to the coordinates instead of going directly to specific row and search only there.